clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc

```diff
@@ -239,9 +243,9 @@ public:
   }
   device_vector &operator=(device_vector &&other) {
     // Move assignment operator:
-    this->_size = std::move(other._size);
-    this->_capacity = std::move(other._capacity);
-    this->_storage = std::move(other._storage);
+    // Swap resource with `other`, so that the old resource owned by `this` will
+    // be deallocated by destructor of `other`.
+    this->swap(other);
```

Comment:
The old resource should be deallocated inside the assignment, not depends on the destructor of `other`.
Like:
   ```
   device_vector dummy;
   this->swap(dummy);
   ```

Reply:
Directly swapping makes move assignment operator exception-free: `std::swap` doesn't throw. The drawback is that the lifetime of old resource may be unexpectedly extended, however, it's still a valid implementation: https://en.cppreference.com/w/cpp/language/move_assignment

> Move assignment operators typically "steal" the resources held by the argument (e.g. pointers to dynamically-allocated objects, file descriptors, TCP sockets, I/O streams, running threads, etc.), rather than make copies of them, and **leave the argument in some valid but otherwise indeterminate state**.

Deallocating resources in the move assignment operator may throw, but the lifetime of old resource is released in-time (intuitively).

There're debates over the move assignment implementations:
- https://stackoverflow.com/questions/6687388/why-do-some-people-use-swap-for-move-assignments
- https://stackoverflow.com/questions/32234623/using-swap-to-implement-move-assignment
- https://stackoverflow.com/questions/9847860/questions-about-the-move-assignment-operator

So I'm okay with both.